File: python/benchmark.py
"""
Benchmark: Python/PyPSA — DC Power Flow & LOPF.
Run alongside julia/benchmarks/benchmark.jl for comparison.

Statistics: each size is measured over SEEDS random topologies (identical seeds
and generator logic as the Julia side), runs_per_seed timed solves each after a
warm-up. CSV: time_ms = mean, plus std_ms, min_ms, n_samples.
"""
import os
for _v in ("OPENBLAS_NUM_THREADS", "OMP_NUM_THREADS", "MKL_NUM_THREADS"):
    os.environ[_v] = "1"          # fair single-thread linear algebra (set before numpy)
import numpy as np
import pypsa
import time
import statistics
import csv
import warnings
import logging

warnings.filterwarnings("ignore")
logging.disable(logging.CRITICAL)

SEEDS = [1, 2, 3]


def generate_network(n_buses, seed=42):
    rng = np.random.default_rng(seed)

    lines = []
    for i in range(1, n_buses):
        x = 0.05 + rng.random() * 0.45
        lines.append((i, i + 1, x))
    for _ in range(max(1, n_buses // 3)):
        u = int(rng.integers(1, n_buses))
        v = int(rng.integers(u + 1, n_buses + 1))
        x = 0.05 + rng.random() * 0.45
        lines.append((u, v, x))

    loads = {}
    total_load = 0.0
    for bus in range(2, n_buses + 1):
        if rng.random() > 0.3:
            p = 50.0 + rng.random() * 450.0
            loads[bus] = p
            total_load += p
    if not loads:
        loads[2] = 200.0
        total_load = 200.0

    generators = {1: total_load * 1.1}
    for bus in range(2, n_buses + 1, 4):
        generators[bus] = loads.get(bus, 0.0) * 0.5 + 50.0

    return n_buses, lines, generators, loads


def build_pypsa_network(n_buses, lines, generators, loads, line_capacity=1e6):
    n = pypsa.Network()
    for bus in range(1, n_buses + 1):
        n.add("Bus", f"Bus{bus}", v_nom=380.0)
    for idx, (f, t, x) in enumerate(lines):
        n.add("Line", f"L{idx}", bus0=f"Bus{f}", bus1=f"Bus{t}",
              x=x, r=0.01, s_nom=line_capacity)
    for bus, p_max in generators.items():
        ctrl = "Slack" if bus == 1 else "PQ"
        n.add("Generator", f"G{bus}", bus=f"Bus{bus}",
              p_nom=p_max, marginal_cost=20.0, control=ctrl)
    for bus, p in loads.items():
        n.add("Load", f"Load{bus}", bus=f"Bus{bus}", p_set=p)
    return n


def bench_stats(run_factory, runs_per_seed):
    """run_factory(seed) -> zero-arg callable. Pool samples over SEEDS × runs."""
    samples = []
    for s in SEEDS:
        run = run_factory(s)
        run()                                   # warm-up
        for _ in range(runs_per_seed):
            t0 = time.perf_counter()
            run()
            samples.append(time.perf_counter() - t0)
    mean = statistics.mean(samples) * 1000
    std = (statistics.stdev(samples) if len(samples) > 1 else 0.0) * 1000
    return mean, std, min(samples) * 1000, len(samples)


print("=" * 74)
print("BENCHMARK: Python/PyPSA — DC Power Flow & LOPF")
print(f"Seeds: {SEEDS}  |  mean +/- std over pooled samples")
print("=" * 74)

DC_SIZES   = [3, 10, 50, 100, 300]
LOPF_SIZES = [3, 10, 50, 100, 300]

print("\n[DC POWER FLOW]  net.lpf()")
print("-" * 74)
print(f"{'Buses':<8} {'Mean (ms)':>12} {'Std (ms)':>12} {'Min (ms)':>12} {'N':>6}")
print("-" * 74)

dc_results = {}
for n in DC_SIZES:
    runs = 10 if n <= 100 else (8 if n <= 500 else (5 if n <= 1000 else 3))

    def factory(s, n=n):
        net = build_pypsa_network(*generate_network(n, seed=s))
        return lambda: net.lpf()

    mean, sd, mn, ns = bench_stats(factory, runs)
    dc_results[n] = (mean, sd, mn, ns)
    print(f"{n:<8} {mean:>12.4f} {sd:>12.4f} {mn:>12.4f} {ns:>6}")

print("\n[LOPF]  net.optimize(solver='highs')")
print("-" * 74)
print(f"{'Buses':<8} {'Mean (ms)':>12} {'Std (ms)':>12} {'Min (ms)':>12} {'N':>6}")
print("-" * 74)

lopf_results = {}
for n in LOPF_SIZES:
    runs = 10 if n <= 50 else (8 if n <= 100 else 4)

    def factory(s, n=n):
        # Build the network ONCE (outside the timed call), matching the Julia
        # side, so the timing measures only optimisation-model assembly + solve,
        # not PyPSA network construction. optimize() rebuilds the linopy model
        # internally each call, just as Julia's solve() rebuilds the JuMP model.
        net = build_pypsa_network(*generate_network(n, seed=s))
        return lambda: net.optimize(solver_name="highs", solver_options={"threads": 1})

    mean, sd, mn, ns = bench_stats(factory, runs)
    lopf_results[n] = (mean, sd, mn, ns)
    print(f"{n:<8} {mean:>12.4f} {sd:>12.4f} {mn:>12.4f} {ns:>6}")

with open("results/python_benchmark.csv", "w", newline="") as f:
    writer = csv.writer(f)
    writer.writerow(["module", "n_buses", "time_ms", "std_ms", "min_ms", "n_samples"])
    for n in DC_SIZES:
        writer.writerow(["DC_PF", n, *dc_results[n]])
    for n in LOPF_SIZES:
        writer.writerow(["LOPF", n, *lopf_results[n]])

print("\n[OK] results/python_benchmark.csv")

File: julia/benchmarks/benchmark.jl
"""
Benchmark: Julia — DC Power Flow & LOPF, via the PowerFlowJulia LIBRARY API.

Benchmarks the *exported* `pf(net)` / `optimize(net)` entry points — exactly the
API the thesis describes — NOT a hand-written kernel, so the Julia-vs-PyPSA
comparison is apples-to-apples (both build a network through their public API
and call their public solver, both LP via HiGHS).

Statistics: each size is measured over SEEDS random topologies (identical to
python/benchmark.py), with `runs_per_seed` timed solves each after a JIT
warm-up. Reported: mean ± std and min over the pooled samples. The CSV keeps
`time_ms` = mean for backward compatibility, plus `std_ms`, `min_ms`.
"""

using Random
using Printf
using Statistics
using LinearAlgebra

BLAS.set_num_threads(1)   # fair single-thread linear algebra (DC PF solve)

include(joinpath(@__DIR__, "..", "src", "PowerFlowJulia.jl"))
using .PowerFlowJulia

const SEEDS = [1, 2, 3]

# ── Network generator (identical to python/benchmark.py) ─────────────────────
function generate_network(n_buses; seed=42)
    rng = MersenneTwister(seed)

    lines = Tuple{Int,Int,Float64}[]
    for i in 1:n_buses-1
        x = 0.05 + rand(rng) * 0.45
        push!(lines, (i, i+1, x))
    end
    for _ in 1:max(1, n_buses ÷ 3)
        u = rand(rng, 1:n_buses-1)
        v = rand(rng, u+1:n_buses)
        x = 0.05 + rand(rng) * 0.45
        push!(lines, (u, v, x))
    end

    loads      = Dict{Int,Float64}()
    total_load = 0.0
    for bus in 2:n_buses
        if rand(rng) > 0.3
            p = 50.0 + rand(rng) * 450.0
            loads[bus] = p
            total_load += p
        end
    end
    isempty(loads) && (loads[2] = 200.0; total_load = 200.0)

    generators = Dict{Int,Float64}(1 => total_load * 1.1)
    for bus in 2:4:n_buses
        generators[bus] = get(loads, bus, 0.0) * 0.5 + 50.0
    end

    return n_buses, lines, generators, loads
end

# ── Build a PowerFlowJulia Network from the generated topology ───────────────
function build_network(n_buses, lines, generators, loads)
    net = Network(baseMVA=100.0)
    for b in 1:n_buses
        add!(net, "Bus", "Bus$b"; v_nom=380.0, slack=(b == 1))
    end
    for (i, (f, t, x)) in enumerate(lines)
        add!(net, "Line", "L$i"; bus0="Bus$f", bus1="Bus$t", x=x, r=0.01, s_nom=1e6)
    end
    for (bus, p_nom) in generators
        add!(net, "Generator", "G$bus"; bus="Bus$bus", p_nom=p_nom, marginal_cost=20.0)
    end
    for (bus, p) in loads
        add!(net, "Load", "D$bus"; bus="Bus$bus", p_set=p)
    end
    return net
end

# Collect timing samples over SEEDS × runs_per_seed; return mean/std/min (ms).
function bench_stats(build, solve, runs_per_seed)
    samples = Float64[]
    for s in SEEDS
        net = build(s)
        solve(net)                              # JIT / warm-up
        for _ in 1:runs_per_seed
            push!(samples, @elapsed solve(net))
        end
    end
    n = length(samples)
    return (mean = mean(samples) * 1000,
            std  = (n > 1 ? std(samples) : 0.0) * 1000,
            min  = minimum(samples) * 1000,
            n    = n)
end

println("="^74)
println("BENCHMARK: Julia (PowerFlowJulia API) — DC Power Flow & LOPF")
println("Seeds: $SEEDS  |  mean ± std over pooled samples")
println("="^74)

DC_SIZES   = [3, 10, 50, 100, 300]
LOPF_SIZES = [3, 10, 50, 100, 300]

println("\n[DC POWER FLOW]  pf(net; method=:dc)")
println("-"^74)
@printf("%-8s %12s %12s %12s %6s\n", "Buses", "Mean (ms)", "Std (ms)", "Min (ms)", "N")
println("-"^74)

dc = Dict{Int,NamedTuple}()
for n in DC_SIZES
    runs = n <= 100 ? 10 : (n <= 500 ? 8 : (n <= 1000 ? 5 : 3))
    st = bench_stats(s -> build_network(generate_network(n, seed=s)...),
                     net -> pf(net; method=:dc, verbose=false), runs)
    dc[n] = st
    @printf("%-8d %12.4f %12.4f %12.4f %6d\n", n, st.mean, st.std, st.min, st.n)
end

println("\n[LOPF]  optimize(net; method=:lopf)   (JuMP + HiGHS)")
println("-"^74)
@printf("%-8s %12s %12s %12s %6s\n", "Buses", "Mean (ms)", "Std (ms)", "Min (ms)", "N")
println("-"^74)

lp = Dict{Int,NamedTuple}()
for n in LOPF_SIZES
    runs = n <= 50 ? 10 : (n <= 100 ? 8 : 4)
    st = bench_stats(s -> build_network(generate_network(n, seed=s)...),
                     net -> optimize(net; method=:lopf, verbose=false), runs)
    lp[n] = st
    @printf("%-8d %12.4f %12.4f %12.4f %6d\n", n, st.mean, st.std, st.min, st.n)
end

open(joinpath(@__DIR__, "..", "..", "results", "benchmarks", "julia_benchmark.csv"), "w") do io
    println(io, "module,n_buses,time_ms,std_ms,min_ms,n_samples")
    for n in DC_SIZES;   s=dc[n]; println(io, "DC_PF,$n,$(s.mean),$(s.std),$(s.min),$(s.n)"); end
    for n in LOPF_SIZES; s=lp[n]; println(io, "LOPF,$n,$(s.mean),$(s.std),$(s.min),$(s.n)"); end
end

println("\n[OK] results/benchmarks/julia_benchmark.csv  |  run python/benchmark.py for PyPSA")
